core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -180,38 +180,48 @@ class GroupCoordinator(val brokerId: Int,
-            // New heartbeat shall be scheduled with new member id.
-            completeAndScheduleNextHeartbeatExpiration(group, oldMember)
-
-            responseCallback(JoinGroupResult(
-              members = if (group.isLeader(newMemberId)) {
-                group.currentMemberMetadata
-              } else {
-                List.empty
-              },
-              memberId = newMemberId,
-              generationId = group.generationId,
-              subProtocol = group.protocolOrNull,
-              leaderId = group.leaderOrNull,
-              error = Errors.NONE))
-          } else {
-            val knownStaticMember = group.get(oldMemberId)
-            updateMemberAndRebalance(group, knownStaticMember, protocols, responseCallback)
+          group.currentState match {
+            case Stable =>
+              info(s"Static member $groupInstanceId with unknown member id rejoins group ${group.groupId} " +
+                s"in ${group.currentState} state. Assigning new member id $newMemberId, while old member $oldMemberId " +
+                "will be removed. No rebalance will be triggered.")
+
+              val oldMember = group.replaceGroupInstance(oldMemberId, newMemberId, groupInstanceId)
+
+              // Heartbeat of old member id will expire without affection since the group no longer contains that member id.
+              // New heartbeat shall be scheduled with new member id.
+              completeAndScheduleNextHeartbeatExpiration(group, oldMember)
+
+              responseCallback(JoinGroupResult(
+                members = if (group.isLeader(newMemberId)) {
+                  group.currentMemberMetadata
+                } else {
+                  List.empty
+                },
+                memberId = newMemberId,
+                generationId = group.generationId,
+                subProtocol = group.protocolOrNull,
+                leaderId = group.leaderOrNull,
+                error = Errors.NONE))
+
+            case _ =>
+              info(s"Static member $groupInstanceId with unkonwn member id rejoins group ${group.groupId} " +
+                s"in ${group.currentState} state. Update its membership with the pre-registered old member id $oldMemberId.")
+
+              val knownStaticMember = group.get(oldMemberId)
+              updateMemberAndRebalance(group, knownStaticMember, protocols, responseCallback)
           }
         } else if (requireKnownMemberId) {
             // If member id required (dynamic membership), register the member in the pending member list
             // and send back a response to call for another join group request with allocated member id.
+          debug(s"Dynamic member with unknown member id rejoins group ${group.groupId} in " +
+              s"${group.currentState} state. Created a new member id $newMemberId and request the member to rejoin with this id.")
             group.addPendingMember(newMemberId)
```

Comment:
Indenting doesn't look right here.

Reply:
It should be 2 spaces in Scala, I think the indenting was wrong before this PR (the indenting of this PR is actually right, but just make it convienent to find out the inconsistency). I can piggy back the re-indenting.